keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1425 @@
+phase: https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/kubelet/status/status_manager.go#L1060-L1068.
+When none of the above conditions is satisfied to compute the finish time we
+fallback to the pod's creation time.
+
+This fix can be considered a preparatory PR before the KEP, as to some extent
+is solves the preexisting issue.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+We introduce a new Job API field, called `.spec.backoffLimitPerIndex`.
+When set it limits the number of retries, counted independently for all indexes.
+
+Additionally, we propose the `.spec.maxFailedIndexes` to control
+the maximal number of failed indexes. Once the number is exceeded the entire
+Job is marked Failed and its execution is terminated.
+
+We also propose to extend the PodFailurePolicy with a new action, called
+`FailIndex` to allow an index to fail fast before reaching the backoff limit
+per index.
+
+### Job API
+
+```golang
+
+// PodFailurePolicyAction specifies how a Pod failure is handled.
+// +enum
+type PodFailurePolicyAction string
+
+const (
+  // This is an action which might be taken on a pod failure - mark the
+  // Job's index as failed to avoid restarts within this index. This action
+  // can only be used when backoffLimitPerIndex is set.
+  PodFailurePolicyActionFailIndex PodFailurePolicyAction = "FailIndex"
+  ...
+)
+...
+
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
+  ...
+  // Specifies the limit for the number of retries within an
+  // index before marking this index as failed. When enabled the number of
+  // failures per index is kept in the pod's
+  // batch.kubernetes.io/job-index-failure-count annotation. It can only
+  // be set when Job's completionMode=Indexed, and the Pod's restart
+  // policy is Never.
+  // When specified, then the backoffLimit field is defaulted to the max int32
+  // value. Also, once specified it cannot be set to nil.
+  // +optional
+  BackoffLimitPerIndex *int32
+
+  // Specifies the maximal number of failed indexes before marking the Job as
+  // failed, when backoffLimitPerIndex is set. Once the number of failed
+  // indexes exceeds this number the entire Job is marked as Failed and its
+  // execution is terminated. When left as nil the job continues execution of
+  // all of its indexes and is marked with the `Complete` Job condition.
+  // It can only be specified when backoffLimitPerIndex is set.
+  // The value is might be up to 10^5 when completions is <= 10^5, or 10^4 when
+  // completions > 10^5.
+  // It defaults to 10^4 when completions is greater than 10^5.
```

Comment:
+1: require users to set the value, instead of defaulting based on another field.

Reply:
I see, I see two options to always either set it:
1. always default to completions (will be rejected if completions > 10^5 and maxFailedIndexes is unspecified)
2. leave it unspecified (if unspecified it is unlimited), but require to be specified when completions > 10^5, then only 10^4 is allowed.

Both sound sensible to me.

I will tentatively update to (2.) as it allows for easier integration with elastic jobs (https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3715-elastic-indexed-job. It will not require moving maxFailedIndexes along with completions and parallelism.

Reply:
2 for me